lib/srv/db/mysql/protocol/version.go

```diff
@@ -0,0 +1,130 @@
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+
+*/
+
+package protocol
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"net"
+
+	"github.com/gravitational/trace"
+	"github.com/siddontang/go-mysql/client"
+	"github.com/siddontang/go-mysql/mysql"
+	mysqlpacket "github.com/siddontang/go-mysql/packet"
+)
+
+// FetchMySQLVersionInternal connects to a MySQL instance with provided dialer and tries to read the server
+// version from initial handshake message. Error is returned in case of connection failure or when MySQL
+// returns ERR package.
+func FetchMySQLVersionInternal(ctx context.Context, dialer client.Dialer, databaseURI string) (string, error) {
+	conn, err := dialer(ctx, "tcp", databaseURI)
+	if err != nil {
+		return "", trace.ConnectionProblem(err, "failed to connect to MySQL")
+	}
+	defer conn.Close()
+
+	// Set connection deadline if passed context has it.
+	if deadline, ok := ctx.Deadline(); ok {
+		if err := conn.SetReadDeadline(deadline); err != nil {
+			return "", trace.Wrap(err)
+		}
+	}
+
+	connBuf := newBufferedConn(ctx, conn)
+	pkgType, err := connBuf.Peek(5)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	// ref: https://dev.mysql.com/doc/internals/en/mysql-packet.html
+	//      https://dev.mysql.com/doc/internals/en/packet-ERR_Packet.html
+	if pkgType[4] == mysql.ERR_HEADER {
+		return readHandshakeError(connBuf)
+	}
+
+	return readHandshakeServerVersion(connBuf)
+}
+
+// readHandshakeServerVersion reads MySQL initial handshake message and returns the server version.
+func readHandshakeServerVersion(connBuf net.Conn) (string, error) {
+	dbConn := mysqlpacket.NewTLSConn(connBuf)
+
+	handshake, err := dbConn.ReadPacket()
+	if err != nil {
+		return "", trace.ConnectionProblem(err, "failed to read the MySQL handshake")
+	}
+
+	// ref: https://dev.mysql.com/doc/internals/en/connection-phase-packets.html#packet-Protocol::Handshake
+	versionLength := bytes.IndexByte(handshake[1:], 0x00)
+	if versionLength == -1 {
+		return "", trace.Errorf("failed to read the MySQL server version")
+	}
+
+	return string(handshake[1 : 1+versionLength]), nil
+}
+
+// readHandshakeError reads and returns an error message from
+func readHandshakeError(connBuf io.Reader) (string, error) {
+	handshakePackage, err := ParsePacket(connBuf)
+	if err != nil {
+		return "", err
+	}
+	errPackage, ok := handshakePackage.(*Error)
+	if !ok {
+		return "", trace.BadParameter("expected MySQL error package, got %T", handshakePackage)
+	}
+	return "", trace.ConnectionProblem(errors.New("failed to fetch MySQL version"), errPackage.Error())
+}
+
+// bufferedConn is a net.Conn wrapper with additional Peek() method.
```

Comment:
```suggestion
// connReader is a net.Conn wrapper with additional Peek() method.
```
Same for the `newXXX` function